notion-server/src/ethereum.rs

```diff
@@ -1,4 +1,77 @@
+use base64;
 use std::collections::HashMap;
+use std::convert::TryFrom;
+use snafu::{Snafu, ResultExt, Backtrace};
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    Decode {
+        backtrace: Backtrace,
+        source: base64::DecodeError,
+    },
+    OutOfBounds {
+        message: String,
+    }
+}
+
+pub mod interface_args {
+    #[derive(Debug, Default)]
+    pub struct GeneralSimulationState {
+        pub num_execution_environments: u32,
+        pub num_shard_chains: u32,
+    }
+
+    #[derive(Debug, Default)]
+    pub struct ExecutionEnvironment {
+        pub base64_encoded_wasm_code: String,
+    }
+
+    impl From<&super::ExecutionEnvironment> for ExecutionEnvironment {
```

Comment:
I love the use of Rust's standard conversion traits! That's awesome.

If it's at all possible, taking a `super::ExecutionEnvironment` instead of a reference would be arguably better. Using a reference means that every conversion from an `args::ExecutionEnvironment` to a `super::ExecutionEnvironment` will copy memory. Using the structs directly means that less allocations happen overall.

In this case, I think that because you're converting from base64 `String` to byte arrays, you'll end up with allocations anyway. If you decide to apply [this comment](#r350313031), you should probably do this one too, otherwise it's less important.

Reply:
I think we *want* to use a reference here, though I'd be happy to discuss more and learn if maybe I'm missing something.

These are 2 different objects (the external interface representation and the internal "true" representation), so allocating / creating a new object seems like what I'd expect to happen in a From statement like this.  Seems like making it force moves would be a more restrictive API, and could be a premature optimization (ie. I don't think this simulation will ever need to worry about allocating slightly more memory than normal).  If performance is or will become a big concern, in that case maybe this is more important to do without using a reference as input.

Agreed with your main point that if we just move over variables instead of allocating new ones that this would be more performance (eg. if we decide to change the interface_args to use Vec<u8> we could just move those over vs creating from scratch).

Reply:
> allocating / creating a new object seems like what I'd expect to happen in a From statement like this

Yeah, allocating the internal struct will happen either way. I'm more concerned about the `String`s and `Vec`s inside.

> Seems like making it force moves would be a more restrictive API

If you pass-by-move, you give the caller the choice between making copies or not. If you pass by reference, you always make a copy, even if you don't need to.

Here's how pass-by-move gives the caller the choice:

```rust
fn pass_by_move_cpy() {
    let obj = SomeBigObject::new();
    led jbo = OtherBigObject::from(obj.clone()); // The caller decides to make a copy.

    // The caller needs obj here, for whatever reason.
    println!("{:?}", obj);
}

fn pass_by_move_nocpy() {
    let obj = SomeBigObject::new();
    led jbo = OtherBigObject::from(obj);
    // The caller doesn't need obj any longer, so moving it save allocations.
}
```

If you use a `From<&Foo>`, it isn't possible to implement `pass_by_move_nocpy`.

This entire discussion is pretty much moot here, since the code has to convert from a `String` to a `Vec` anyway. Making this change only makes sense if you apply [this suggestion](https://github.com/quilt/simulation/pull/14#discussion_r350313031).

Reply:
> premature optimization

Absolutely! But it's a good opportunity to discuss API design in Rust ;3

Reply:
I think for now it will be simpler to keep this as-is with references, as it makes a number of things easier -- no need to clone internal structs, it allows the use of the `map` function on `vec`, etc. If we feel strongly about changing this later, we could do that then.  Also, at least for the return values (converting from internal to interface structs), we'd pretty much always want to re-allocate, since we don't want to change ownership of the internal structs.

So, unless you feel strongly, I'd prefer to leave it as-is for now, but if you feel strongly let me know and I'm cool with updating it.

I do agree that having it be `move` by default leaves the decision to `clone` to the caller of the function. This is a small distinction, but worth noting that there could also be some cases where the result of the `from` function returns a subset of the fields of the original struct, that could be less efficient because the `clone` would re-allocate all fields instead of just the ones necessary.

Reply:
> it allows the use of the map function on vec

You're looking for `into_iter` I think.

The rest are great points. I'm fine either way :3